### PR TITLE
Code experiment: apply facet pattern.

### PR DESCRIPTION
**Pull request by brightbyte**
The idea of "facets" as implemented here is also known as the "role object" or "extension" pattern.

This implementation uses a helper object, FacetManager, that is used via composition to provide facet support to other objects that implement FacetContainer.

Facet support is added to the following classes: Term, SiteLink, Statement, and SnakObject.

The following classes are converted to facets: DerivedPropertyValueSnak and TermFallback.

Performance implications were not considered for this experiment. The idea is to provide sample code as a basis for discussion.

[Bug: T118860](http://phabricator.wikimedia.org/T118860)

### CODE REVIEW

File: src/SiteLink.php [hunk 1]
```diff
@@ -38,6 +42,11 @@ class SiteLink implements Comparable {
 	private $badges;
 
 	/**
+	 * @var FacetManager
+	 */
+	private $facetManager;
```
Comment (Benestar) [inline src/SiteLink.php]: Where would $facetManager be set to some value? As far as I see it will always be null and not usable in the current implementation of the methods. When would this object be instantiated?

File: RELEASE-NOTES.md [hunk 2]
```diff
@@ -1,7 +1,11 @@
 # Wikibase DataModel release notes
 
-## Version 4.4.0 (dev)
-
+## Version 5 (dev)
```
Comment (JeroenDeDauw) [inline RELEASE-NOTES.md]: No need to make breaking changes. `TermFallback` and `DerivedPropertyValueSnak` can be deprecated and scheduled for removal in the next breaking release.

File: src/Facet/FacetContainer.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	public function getFacet( $name, $type = null );
+
+	/**
+	 * Adds a facet object. Any facet object previously registered under the same name
+	 * is discarded.
+	 *
+	 * @param string $name
+	 * @param object $facetObject The facet object
+	 */
+	public function addFacet( $name, $facetObject );
```
Comment (Jonas-WMDE) [inline src/Facet/FacetContainer.php]: Why not name it setFacet as there can only be one facet under the same name?

Comment (brightbyte) [inline src/Facet/FacetContainer.php]: no strong feelings about that

**Comment by JeroenDeDauw** [comment]
No strong objection against the concept, though no endorsement either, as I am not convinced an approach such as this makes sense.

Of course before any such thing is merged it needs to be reviewed beyond the conceptual level, which I have not yet done so. Please indicate when the time to proceed with that has arrived.

File: src/Facet/FacetContainer.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	 *
+	 * @return string[]
+	 */
+	public function listFacets();
+
+	/**
+	 * Returns a facet object.
+	 *
+	 * @param string $name
+	 * @param string|null $type The desired type
```
Comment (thiemowmde) [inline src/Facet/FacetContainer.php]: 1. What "type"? We are using this term for a lot of different things. I honestly have no idea which one is meant here.
2. "Desired" how?
3. What is supposed to happen when this is null?

File: src/Facet/FacetContainer.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	 *
+	 * @throws NoSuchFacetException if no facet object is found for $name
+	 * @throws MismatchingFacetException if the facet object is not compatible with $type
+	 * @return object
+	 */
+	public function getFacet( $name, $type = null );
+
+	/**
+	 * Adds a facet object. Any facet object previously registered under the same name
+	 * is discarded.
```
Comment (thiemowmde) [inline src/Facet/FacetContainer.php]: Should be a setter then.

File: src/Facet/FacetContainer.php (hunk 3 above)
Comment (thiemowmde) [inline src/Facet/FacetContainer.php]: Type hint against `object`?

File: src/Facet/FacetContainer.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	 * @param string|null $type The desired type
+	 *
+	 * @throws NoSuchFacetException if no facet object is found for $name
+	 * @throws MismatchingFacetException if the facet object is not compatible with $type
+	 * @return object
+	 */
+	public function getFacet( $name, $type = null );
+
+	/**
+	 * Adds a facet object. Any facet object previously registered under the same name
```
Comment (thiemowmde) [inline src/Facet/FacetContainer.php]: I wonder why this interface does have a setter/adder at all? I expected the knowledge which facets are supported, how they are build and so on to be an implementation detail of the class that implements the interface. But with this public (!) setter everybody is free to add whatever object they want and call it a "facet". Why is this needed? Can't this be protected or omitted?

File: src/Facet/MismatchingFacetException.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @licence GNU GPL v2+
+ * @author Daniel Kinzler
+ */
+class MismatchingFacetException extends RuntimeException {
+
+	/**
+	 * @param string $facetName
+	 */
+	public function __construct( $facetName, $expectedType ) {
```
Comment (thiemowmde) [inline src/Facet/MismatchingFacetException.php]: Type doc missing.

Can the type be anything but a string?

File: src/Internal/FacetManager.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	 * @return object
+	 */
+	public function getFacet( $name, $type = null ) {
+		if ( !isset( $this->facets[$name] ) ) {
+			throw new NoSuchFacetException( $name );
+		}
+
+		$facet = $this->facets[$name];
+
+		// TODO: if $facet is callable, call it, and replace $this->facets[$name] with the result.
```
Comment (thiemowmde) [inline src/Internal/FacetManager.php]: This is the first time I'm reading something about facets being callable. I assume this targets deferred deserialization. My issue with this comment is that it describes a very specific implementation. I suggest to leave this out as long as this implementation is not discussed.

File: src/Facet/FacetContainer.php (hunk 6 above)
Comment (Benestar) [inline src/Facet/FacetContainer.php]: @thiemowmde as far as I understand the facets should be added by services outside of this class/component and thus this setter needs to be public. However, I'm also concerned where we define which factes are available for which model. Currently this is flexible but also quite arbitrary.

File: src/Snak/DerivedValues.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Wikibase\DataModel\Snak;
+
+use DataValues\DataValue;
+use InvalidArgumentException;
+
+/**
+ * Value objects for managing a set of derived DataValues.
```
Comment (thiemowmde) [inline src/Snak/DerivedValues.php]: Imutable.

File: src/Snak/SnakObject.php [hunk 10]
```diff
@@ -116,4 +125,46 @@ public function unserialize( $serialized ) {
...
+	/**
+	 * @param string $name
+	 * @param object $facetObject
+	 */
+	public function addFacet( $name, $facetObject ) {
+		if ( !$this->facetManager ) {
+			$this->facetManager = new FacetManager();
+		}
+
+		$this->facetManager->addFacet( $name, $facetObject );
```
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: The whole implementation here is now copy-pasted several times, in several different classes. To me this screams "abstract base class!", but I do not know if we can agree on this.

File: src/Statement/Statement.php [hunk 11]
```diff
@@ -298,4 +307,46 @@ public function equals( $target ) {
...
+	/**
+	 * @param string $name
+	 * @param object $facetObject
+	 */
+	public function addFacet( $name, $facetObject ) {
+		if ( !$this->facetManager ) {
+			$this->facetManager = new FacetManager();
+		}
+
+		$this->facetManager->addFacet( $name, $facetObject );
```
Comment (thiemowmde) [inline src/Statement/Statement.php]: Here it is the third time.

File: src/Term/Term.php [hunk 12]
```diff
@@ -77,4 +86,46 @@ public function equals( $target ) {
...
+	/**
+	 * @param string $name
+	 * @param object $facetObject
+	 */
+	public function addFacet( $name, $facetObject ) {
+		if ( !$this->facetManager ) {
+			$this->facetManager = new FacetManager();
+		}
+
+		$this->facetManager->addFacet( $name, $facetObject );
```
Comment (thiemowmde) [inline src/Term/Term.php]: And again.

File: src/Snak/SnakObject.php (hunk 10 above)
Comment (Benestar) [inline src/Snak/SnakObject.php]: Would this be a valid use case for [traits](http://php.net/manual/de/language.oop5.traits.php)?

File: src/Facet/FacetContainer.php [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	 * @return boolean
+	 */
+	public function hasFacet( $name );
+
+	/**
+	 * Returns a list of supported facet names.
+	 *
+	 * @return string[]
+	 */
+	public function listFacets();
```
Comment (adrianheine) [inline src/Facet/FacetContainer.php]: What would this be used for?

File: src/Snak/SnakObject.php (hunk 10 above)
Comment (thiemowmde) [inline src/Snak/SnakObject.php]: We need to switch to PHP 5.4 first before being able to use traits. And I remember hearing opinions that we should probably never use traits, because they do not solve the issues abstract base classes introduce.

**Comment by JeroenDeDauw** [comment]
Since this is an experiment from 2015 with no updates since I'm assuming this can be closed.
